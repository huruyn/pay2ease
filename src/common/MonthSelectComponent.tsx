import { Form } from 'react-bootstrap';

export default function MonthSelectComponent({
  onChange,
  prefix = 'option',
  currentDate,
}: any) {
  const getDateAry = () => {
    const result: any = [];
    let prevYearValue = '';
    for (let i = 0; i < 12; i += 1) {
      if (currentDate) {
        const calculatedDate = currentDate.minus({ months: i });
        const dateStr = calculatedDate.toFormat('MMMM yyyy');
        const yearValue = calculatedDate.toFormat('yyyy');
        if (prevYearValue !== yearValue) {
          result.push({
            label: yearValue,
            items: [],
          });
          prevYearValue = yearValue;
        }
        const item = {
          value: dateStr,
        };
        result[result.length - 1].items.push(item);
      }
    }
    return result;
  };

  return (
    <Form.Select style={{ maxWidth: 300 }} onChange={onChange} className="month-select-comp">
      {getDateAry().map((item: any) => (
        <optgroup label={`${item.label}`} key={`${prefix}_${item.label}`}>
          {item.items.map((dateValue: any) => (
            <option value={dateValue.value} key={`${prefix}_${dateValue.value}`}>
              {dateValue.value}
            </option>
          ))}
        </optgroup>
      ))}
    </Form.Select>
  );
}
