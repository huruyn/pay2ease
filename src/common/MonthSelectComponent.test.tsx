import { render, fireEvent } from '@testing-library/react';
import { DateTime } from 'luxon';
import MonthSelectComponent from './MonthSelectComponent';

const ReactTestRenderer = require('react-test-renderer');

describe('MonthSelectComponent Component', () => {
  it('matches the snapshot', () => {
    const tree = ReactTestRenderer.create(
      <MonthSelectComponent currentDate={DateTime.utc()} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render options inside a MonthSelectComponent', () => {
    const { container } = render(
      <MonthSelectComponent currentDate={DateTime.utc()} />,
    );
    const options = container.querySelectorAll('option');
    expect(options).toHaveLength(12);
  });

  it('Click MonthSelectComponent Component', () => {
    const onChange = jest.fn();
    const { container } = render(
      <MonthSelectComponent onChange={onChange} currentDate={DateTime.utc()} />,
    );
    const mEvent = { preventDefault: jest.fn() };
    const optionElements = container.querySelectorAll('.month-select-comp');
    expect(optionElements).toHaveLength(1);
    fireEvent.change(optionElements[0], mEvent);
    expect(onChange).toBeCalledTimes(1);
  });
});
