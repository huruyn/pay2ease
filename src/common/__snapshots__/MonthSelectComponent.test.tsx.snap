// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthSelectComponent Component matches the snapshot 1`] = `
<select
  className="month-select-comp form-select"
  style={
    Object {
      "maxWidth": 300,
    }
  }
>
  <optgroup
    label="2022"
  >
    <option
      value="February 2022"
    >
      February 2022
    </option>
    <option
      value="January 2022"
    >
      January 2022
    </option>
  </optgroup>
  <optgroup
    label="2021"
  >
    <option
      value="December 2021"
    >
      December 2021
    </option>
    <option
      value="November 2021"
    >
      November 2021
    </option>
    <option
      value="October 2021"
    >
      October 2021
    </option>
    <option
      value="September 2021"
    >
      September 2021
    </option>
    <option
      value="August 2021"
    >
      August 2021
    </option>
    <option
      value="July 2021"
    >
      July 2021
    </option>
    <option
      value="June 2021"
    >
      June 2021
    </option>
    <option
      value="May 2021"
    >
      May 2021
    </option>
    <option
      value="April 2021"
    >
      April 2021
    </option>
    <option
      value="March 2021"
    >
      March 2021
    </option>
  </optgroup>
</select>
`;
